In [16]:
import pandas as pd
import matplotlib.pyplot as plt

In [6]:
df = pd.read_csv('data/pcaps.csv')
df.head()

No.      Time    Source Destination Protocol  Length  \
0    1  0.000000  10.0.1.6    10.0.0.4      DNS      81   
1    2  0.001528  10.0.0.4    10.0.1.6      DNS      97   
2    3  0.003326  10.0.1.6    10.0.0.4      TCP      66   
3    4  0.004315  10.0.0.4    10.0.1.6      TCP      66   
4    5  0.004369  10.0.1.6    10.0.0.4      TCP      54   

                                                Info  
0      Standard query 0xa0dc A NEWYORK.dmevals.local  
1  Standard query response 0xa0dc A NEWYORK.dmeva...  
2  56880  >  445 [SYN] Seq=0 Win=64240 Len=0 MSS=...  
3  445  >  56880 [SYN, ACK] Seq=0 Ack=1 Win=65535...  
4   56880  >  445 [ACK] Seq=1 Ack=1 Win=524544 Len=0

In [107]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7284 entries, 0 to 7283
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   No.          7284 non-null   int64  
 1   Time         7284 non-null   float64
 2   Source       7284 non-null   object 
 3   Destination  7284 non-null   object 
 4   Protocol     7284 non-null   object 
 5   Length       7284 non-null   int64  
 6   Info         7284 non-null   object 
dtypes: float64(1), int64(2), object(4)
memory usage: 398.5+ KB


In [20]:
df['Length'].mean()

1131.2689456342669

In [24]:
df.groupby('Source').Source.count()

Source
10.0.0.4        453
10.0.1.4        888
10.0.1.6       3636
192.168.0.4    2307
Name: Source, dtype: int64

## Podejrzane ip - 192.168.0.4

In [26]:
df[df['Source'] == '192.168.0.4']

No.         Time       Source Destination Protocol  Length  \
2211  2212  2044.908808  192.168.0.4    10.0.1.6      TCP      66   
2214  2215  2044.910063  192.168.0.4    10.0.1.6      TCP      54   
2215  2216  2044.912052  192.168.0.4    10.0.1.6  TLSv1.2    1462   
2217  2218  2044.916148  192.168.0.4    10.0.1.6  TLSv1.2     296   
2219  2220  2044.960585  192.168.0.4    10.0.1.6      TCP      54   
...    ...          ...          ...         ...      ...     ...   
7200  7201  2284.926183  192.168.0.4    10.0.1.6      TCP      54   
7201  7202  2284.927902  192.168.0.4    10.0.1.6  TLSv1.2    1462   
7203  7204  2284.931496  192.168.0.4    10.0.1.6  TLSv1.2     296   
7205  7206  2284.933024  192.168.0.4    10.0.1.6  TLSv1.2     196   
7206  7207  2284.933024  192.168.0.4    10.0.1.6  TLSv1.2      85   

                                                   Info  
2211  8443  >  57091 [SYN, ACK] Seq=0 Ack=1 Win=6424...  
2214  8443  >  57091 [ACK] Seq=1 Ack=518 Win=64128 L...  
2215  Server Hello, Certificate, Server Key Exchange...  
2217  New Session Ticket, Change Cipher Spec, Encryp...  
2219  8443  >  57091 [ACK] Seq=1651 Ack=895 Win=6412...  
...                                                 ...  
7200  8443  >  57475 [ACK] Seq=1 Ack=518 Win=64128 L...  
7201  Server Hello, Certificate, Server Key Exchange...  
7203  New Session Ticket, Change Cipher Spec, Encryp...  
7205                                   Application Data  
7206                                    Encrypted Alert  

[2307 rows x 7 columns]

## 192.168.0.4 kompunikuje się z 10.0.16 przy pomocy podejrzanego portu 8443

In [27]:
df[(df['Source'] == '192.168.0.4') & (df['Destination'] == '10.0.1.6')]

No.         Time       Source Destination Protocol  Length  \
2211  2212  2044.908808  192.168.0.4    10.0.1.6      TCP      66   
2214  2215  2044.910063  192.168.0.4    10.0.1.6      TCP      54   
2215  2216  2044.912052  192.168.0.4    10.0.1.6  TLSv1.2    1462   
2217  2218  2044.916148  192.168.0.4    10.0.1.6  TLSv1.2     296   
2219  2220  2044.960585  192.168.0.4    10.0.1.6      TCP      54   
...    ...          ...          ...         ...      ...     ...   
7200  7201  2284.926183  192.168.0.4    10.0.1.6      TCP      54   
7201  7202  2284.927902  192.168.0.4    10.0.1.6  TLSv1.2    1462   
7203  7204  2284.931496  192.168.0.4    10.0.1.6  TLSv1.2     296   
7205  7206  2284.933024  192.168.0.4    10.0.1.6  TLSv1.2     196   
7206  7207  2284.933024  192.168.0.4    10.0.1.6  TLSv1.2      85   

                                                   Info  
2211  8443  >  57091 [SYN, ACK] Seq=0 Ack=1 Win=6424...  
2214  8443  >  57091 [ACK] Seq=1 Ack=518 Win=64128 L...  
2215  Server Hello, Certificate, Server Key Exchange...  
2217  New Session Ticket, Change Cipher Spec, Encryp...  
2219  8443  >  57091 [ACK] Seq=1651 Ack=895 Win=6412...  
...                                                 ...  
7200  8443  >  57475 [ACK] Seq=1 Ack=518 Win=64128 L...  
7201  Server Hello, Certificate, Server Key Exchange...  
7203  New Session Ticket, Change Cipher Spec, Encryp...  
7205                                   Application Data  
7206                                    Encrypted Alert  

[2307 rows x 7 columns]

## 192.168.0.4 rozmawia tylko z 10.0.1.6. Widać to ponieważ powyzsze dwa polecenia zwrocily ta sama liczbe wierszy - 2307. 

In [29]:
df_sus = df[(df['Source'] == '192.168.0.4') & (df['Destination'] == '10.0.1.6')]
df_sus.groupby('Protocol').Source.count()

Protocol
TCP         774
TLSv1.2    1533
Name: Source, dtype: int64

## Komunikacja przebiegala glownie po zaszyfrowanym kanale tls1.2

In [32]:
df_sus['Length'].mean()

791.0814911140009

In [36]:
df_sus['Src port'] = df_sus['Info'].str.extract(r'(8443)')
df_sus

/tmp/ipykernel_27671/2679137617.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sus['Src port'] = df_sus['Info'].str.extract(r'(8443)')


No.         Time       Source Destination Protocol  Length  \
2211  2212  2044.908808  192.168.0.4    10.0.1.6      TCP      66   
2214  2215  2044.910063  192.168.0.4    10.0.1.6      TCP      54   
2215  2216  2044.912052  192.168.0.4    10.0.1.6  TLSv1.2    1462   
2217  2218  2044.916148  192.168.0.4    10.0.1.6  TLSv1.2     296   
2219  2220  2044.960585  192.168.0.4    10.0.1.6      TCP      54   
...    ...          ...          ...         ...      ...     ...   
7200  7201  2284.926183  192.168.0.4    10.0.1.6      TCP      54   
7201  7202  2284.927902  192.168.0.4    10.0.1.6  TLSv1.2    1462   
7203  7204  2284.931496  192.168.0.4    10.0.1.6  TLSv1.2     296   
7205  7206  2284.933024  192.168.0.4    10.0.1.6  TLSv1.2     196   
7206  7207  2284.933024  192.168.0.4    10.0.1.6  TLSv1.2      85   

                                                   Info Src port  
2211  8443  >  57091 [SYN, ACK] Seq=0 Ack=1 Win=6424...     8443  
2214  8443  >  57091 [ACK] Seq=1 Ack=518 Win=64128 L...     8443  
2215  Server Hello, Certificate, Server Key Exchange...      NaN  
2217  New Session Ticket, Change Cipher Spec, Encryp...      NaN  
2219  8443  >  57091 [ACK] Seq=1651 Ack=895 Win=6412...     8443  
...                                                 ...      ...  
7200  8443  >  57475 [ACK] Seq=1 Ack=518 Win=64128 L...     8443  
7201  Server Hello, Certificate, Server Key Exchange...      NaN  
7203  New Session Ticket, Change Cipher Spec, Encryp...      NaN  
7205                                   Application Data      NaN  
7206                                    Encrypted Alert      NaN  

[2307 rows x 8 columns]

In [46]:
df[(df['Destination'] == '192.168.0.4') & (df['Source'] == '10.0.1.6')]

No.         Time    Source  Destination Protocol  Length  \
2210  2211  2044.906584  10.0.1.6  192.168.0.4      TCP      66   
2212  2213  2044.908855  10.0.1.6  192.168.0.4      TCP      54   
2213  2214  2044.909115  10.0.1.6  192.168.0.4  TLSv1.2     571   
2216  2217  2044.915212  10.0.1.6  192.168.0.4  TLSv1.2     180   
2218  2219  2044.916537  10.0.1.6  192.168.0.4  TLSv1.2     305   
...    ...          ...       ...          ...      ...     ...   
7199  7200  2284.925652  10.0.1.6  192.168.0.4  TLSv1.2     571   
7202  7203  2284.930603  10.0.1.6  192.168.0.4  TLSv1.2     180   
7204  7205  2284.931626  10.0.1.6  192.168.0.4  TLSv1.2     570   
7207  7208  2284.933045  10.0.1.6  192.168.0.4      TCP      54   
7208  7209  2284.933241  10.0.1.6  192.168.0.4      TCP      54   

                                                   Info  
2210  57091  >  8443 [SYN] Seq=0 Win=64240 Len=0 MSS...  
2212  57091  >  8443 [ACK] Seq=1 Ack=1 Win=524544 Len=0  
2213                                       Client Hello  
2216  Client Key Exchange, Change Cipher Spec, Encry...  
2218                                   Application Data  
...                                                 ...  
7199                                       Client Hello  
7202  Client Key Exchange, Change Cipher Spec, Encry...  
7204                                   Application Data  
7207  57475  >  8443 [ACK] Seq=1160 Ack=1825 Win=524...  
7208  57475  >  8443 [RST, ACK] Seq=1160 Ack=1825 Wi...  

[2666 rows x 7 columns]

In [48]:
df_sus = df[((df['Destination'] == '192.168.0.4') & (df['Source'] == '10.0.1.6')) | ((df['Source'] == '192.168.0.4') & (df['Destination'] == '10.0.1.6'))]
df_sus

No.         Time       Source  Destination Protocol  Length  \
2210  2211  2044.906584     10.0.1.6  192.168.0.4      TCP      66   
2211  2212  2044.908808  192.168.0.4     10.0.1.6      TCP      66   
2212  2213  2044.908855     10.0.1.6  192.168.0.4      TCP      54   
2213  2214  2044.909115     10.0.1.6  192.168.0.4  TLSv1.2     571   
2214  2215  2044.910063  192.168.0.4     10.0.1.6      TCP      54   
...    ...          ...          ...          ...      ...     ...   
7204  7205  2284.931626     10.0.1.6  192.168.0.4  TLSv1.2     570   
7205  7206  2284.933024  192.168.0.4     10.0.1.6  TLSv1.2     196   
7206  7207  2284.933024  192.168.0.4     10.0.1.6  TLSv1.2      85   
7207  7208  2284.933045     10.0.1.6  192.168.0.4      TCP      54   
7208  7209  2284.933241     10.0.1.6  192.168.0.4      TCP      54   

                                                   Info  
2210  57091  >  8443 [SYN] Seq=0 Win=64240 Len=0 MSS...  
2211  8443  >  57091 [SYN, ACK] Seq=0 Ack=1 Win=6424...  
2212  57091  >  8443 [ACK] Seq=1 Ack=1 Win=524544 Len=0  
2213                                       Client Hello  
2214  8443  >  57091 [ACK] Seq=1 Ack=518 Win=64128 L...  
...                                                 ...  
7204                                   Application Data  
7205                                   Application Data  
7206                                    Encrypted Alert  
7207  57475  >  8443 [ACK] Seq=1160 Ack=1825 Win=524...  
7208  57475  >  8443 [RST, ACK] Seq=1160 Ack=1825 Wi...  

[4973 rows x 7 columns]

In [92]:
df_sus['Src port'] = df_sus['Info'].str.extract(r'(^\d*)')
df_hist = df_sus.groupby('Src port').Source.count()
df_hist

/tmp/ipykernel_27671/616965053.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_sus['Src port'] = df_sus['Info'].str.extract(r'(^\d*)')


Src port
         2661
57091       6
57092       4
57093       4
57094       4
         ... 
57472       4
57473       4
57474       4
57475       4
8443      774
Name: Source, Length: 378, dtype: int64

In [83]:
%matplotlib

Using matplotlib backend: TkAgg


In [98]:
df_hist2 = df_hist.reset_index()
df_hist3 = df_hist2.rename(columns={'Source': 'Count'})
df_hist3

Src port  Count
0              2661
1      57091      6
2      57092      4
3      57093      4
4      57094      4
..       ...    ...
373    57472      4
374    57473      4
375    57474      4
376    57475      4
377     8443    774

[378 rows x 2 columns]

In [102]:
df_hist4 = df_hist3[df_hist3['Count'] > 5]
df_hist4

Src port  Count
0              2661
1      57091      6
12     57102      8
69     57160     23
94     57185     13
377     8443    774

In [104]:
df_hist5 = df_hist4.drop(0)
df_hist5

Src port  Count
1      57091      6
12     57102      8
69     57160     23
94     57185     13
377     8443    774

In [105]:
df_hist5.plot(x='Src port', y='Count', kind='bar')

<AxesSubplot:xlabel='Src port'>

## Jak widać na powyzszym wykresie najwiecej pakietow idzie z portu 8443.
## Na tym etapie jestesmy w stanie przyjac hipotetyczny prog, ktory bedzie nam definiowal zajscie anomalii.
## Rozsadna wartoscia, przy porcie 8443, ktory jest bardzo podejrzany, jest 100 pakietow. 